plugins/mtk.c

```diff
@@ -326,6 +334,142 @@ static void sim_inserted(struct ril_msg *message, gpointer user_data)
 	}
 }
 
+static void set_trm_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct ofono_modem *modem = user_data;
+	struct mtk_data *md = ofono_modem_get_data(modem);
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: RIL error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	g_ril_print_response_no_args(md->ril, message);
+
+	/*
+	 * rild will close now the socket, so we set our state to power off:
+	 * we will set the modem to powered when we reconnect to the socket.
+	 * TODO: Modify reconnect timeout for this case, as the time is much
+	 * less in arale, compared to krillin.
+	 */
+	md->ofono_online = FALSE;
+	ofono_modem_set_powered(md->modem, FALSE);
+}
+
+static void store_type_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct ofono_modem *modem = user_data;
+	struct mtk_data *md = ofono_modem_get_data(modem);
+	struct parcel rilp;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: RIL error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	g_ril_print_response_no_args(md->ril, message);
+
+	/*
+	 * Send SET_TRM, which reloads the FW. We do not know the meaning of the
+	 * magic number 0x0B.
+	 */
+	g_mtk_request_set_trm(md->ril, 0x0B, &rilp);
+
+	if (g_ril_send(md->ril, MTK_RIL_REQUEST_SET_TRM, &rilp,
+			set_trm_cb, modem, NULL) == 0)
```

Comment:
From my Android log dump, there is no MTK_RIL_REQUEST_SET_TRM ever been sent, but there are STORE_MODEM_TYPE ones. It follows that after STORE_MODEM_TYPE responses, it's not necessarily followed by a MTK_RIL_REQUEST_SET_TRM req. Sending this request has no effect on my tests, but the two lines
```C
md->ofono_online = FALSE;
ofono_modem_set_powered(md->modem, FALSE);
```
in the `set_trm_cb` will cause the device cease to register to the network.